sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala

```diff
@@ -54,38 +54,38 @@ object Cast {
 
     case (NullType, _) => true
 
-    case (_, StringType) => true
+    case (_, StringType | CharType(_) | VarcharType(_)) => true
```

Comment:
hey sorry but I can't follow why we should fix this here. char and varchars are not supposed to be directly used in these expressions

Reply:
cc @cloud-fan and @yaooqinn.

Reply:
Do you have an end-to-end test? char/varchar type is carefully handled and they will be replaced with string type at the very beginning. We should not hit char/varchar type.

Reply:
the method cast will change the type char/varchar to StringType,  so   cast method allows conversion from other types to CHAR/VARCHAR, although a warn message is printed.


![image](https://user-images.githubusercontent.com/3105102/128316157-26022c5e-5ac2-471a-aeff-42b0888986e5.png)


![image](https://user-images.githubusercontent.com/3105102/128316118-9dc15053-6a4e-48d0-aeeb-f0f3edf7c42b.png)


in my application code,  I first use the canCast method to determine whether the type conversion can be performed before calling the cast method for conversion。



**i  think the canCast method muse be consistent with Cast method**


@cloud-fan

Reply:
in my application code， I bring this commit, and rebuid and test, all ut and end-to-end test is passed, I'm working on upgrading spark from 2.4.1 to 3.1.1, which Kylin Enterprise Edition depends on   

@cloud-fan

Reply:
If you want to use these private Spark APIs (`Cast` is private), you should also follow these internal assumptions and call `replaceCharVarcharWithStringForCast` before dealing with data types.

I don't see why `Cast.canCast` should match `Column.cast`. They are totally different things. One is a low-level util function which is used internally inside Spark, the other is a public user-facing API.

Reply:
I think people call the Cast.canCast method to use cast for the next step. Since the cast method internally converts from CHAR/VARCHAR type to StringType by default, then I think the canCast method should also convert CHAR/VARCHAR like StringType 

@cloud-fan

Reply:
> I think people call the Cast.canCast method to use cast for the next step.

I don't think this is common. At least this is not recommended. `Cast` is private and if you must use it, please follow the internal assumptions of Spark.

Inside Spark, the input of `Cast.canCast` can never be char/varchar. Please follow this internal assumption if you want to use this private API. I can also add an assert there to make it clearer.

You can call `replaceCharVarcharWithStringForCast` before calling `Cast.canCas`, in your application code.

Reply:
ok

Reply:
I can help to add this assert to prevent misuse by others, if you mean that char/varchar cannot be used in the cancast method

@cloud-fan